### ISSUE TITLE
Change Request: Export file globs of supported files

### ISSUE THREAD

**Issue by kirkwaiblinger** [eslint/eslint#20590]
### ESLint version

10.0.0

### What problem do you want to solve?

Since https://github.com/eslint/eslint/issues/20075 &rarr; https://github.com/eslint/eslint/pull/20158, eslint has recommended using `files` (often in combination with `extends`) on essentially all config objects. That means that, now, a user needs to make a decision on every config object that they write what file glob it should apply to (or whether to leave it unscoped, which is less recommended nowadays). It's not entirely clear what the best choice is. For JS (without TS), a good choice is probably `**/*.{mjs,cjs,js,jsx}`. But, what if I don't use `mjs` files? Should I omit that from the glob? What if I might add `mjs` files to the project at some point in the future? Will I remember to add mjs into the glob? With a user-defined string in each config it becomes a YAGNI vs sensible-forward-thinking question to balance each time one writes a config object.

Especially now that ESLint is a multi-language linter, I propose that ESLint should provide some way to users to match all files are part of a language. This glob could then be used as a default anywhere one is writing a config for that language, rather than going through the cognitive burden of individually specifying file types.

For example, one might use such a utility like so:

```js
import js from '@eslint/js';
import { defineConfig, globs } from 'eslint/config';

export const defineConfig({
  files: [globs.js], // or, globs.jsAndTs if they want TS linted too
  extends: [
    js.configs.recommended
  ],
});
```

Currently, the eslint docs use `**/*.js` as the "default" glob throughout (for example, [Getting Started](https://eslint.org/docs/latest/use/getting-started)). I think this is unfortunate, because it would be preferable to have the standard setup include ESLint being enabled on all JS-language files. Furthermore, it's confusing that by default ESLint _runs_ on `**/*.{cjs,mjs,js}` files, so in one sense the default setup is to _implicitly_ run on `**/*.{cjs,mjs,js}`, and in other sense the default setup is to _explicitly_ run on only `**/*.js` files. This inconsistency increases the decision-making burden in getting started using ESLint.

--

co-authored-by: @JoshuaKGoldberg 

### What do you think is the correct solution?

```js
export const globs = {
  js: "**/*.{mjs,cjs,js,jsx}"
  ts: "**/*.{mts,cts,ts,tsx}"
  jsAndTs: "**/*.{mjs,cjs,js,jsx,mts,cts,ts,tsx}"
}
```

or, potentially a glob per language

```js
export const globs = {
  jsLang: "**/*.{mjs,cjs,js,jsx,mts,cts,ts,tsx}"
  mdLang: "**/*.{md,mdx}",
  jsonLang: "**/*.{json,jsonc}"
}
// ...
```

### Participation

- [x] I am willing to submit a pull request for this change.

### AI acknowledgment

- [x] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

### Additional comments

At typescript-eslint, we have begun working on converting our docs to `files`/`extends` following ESLint's lead (see https://github.com/typescript-eslint/typescript-eslint/issues/12092 &rarr; https://github.com/typescript-eslint/typescript-eslint/pull/12098#discussion_r2886696815), but found that having such globs would be more or less a necessity for writing coherent docs (since the appropriate default glob for typescript-eslint would be `'**/*.{mjs,js,cjs,jsx,mts,ts,cts,tsx}'`). 

See https://github.com/typescript-eslint/typescript-eslint/issues/12102 for essentially the same proposal on the typescript-eslint side. 

---

Related issues of confusion around file inclusion and `files`: https://github.com/eslint/eslint/discussions/20429, https://github.com/eslint/eslint/issues/20354, https://github.com/eslint/eslint/issues/20264, https://github.com/eslint/eslint/issues/19838, https://github.com/typescript-eslint/typescript-eslint/issues/11300

**Comment by kirkwaiblinger** [comment]
I guess, ultimately, there are a few possible solutions if the problem statement is agreed to. Off the top of my head, we at least have:

1. ESLint, the multi-language linting platform, provides globs for matching the languages that the ESLint platform defines/supports.
2. ESLint, the multi-language linting platform, standardizes a recommendation that _plugins_ provide globs to which they are intended apply. In that case, ESLint, the JS plugin for the multi-language linting platform, would provide globs to which it applies (as well as the ESLint MD plugin, etc).

**Comment by snitin315** [comment]
>```js
> export const globs = {
>  jsLang: "**/*.{mjs,cjs,js,jsx,mts,cts,ts,tsx}",
>  mdLang: "**/*.{md,mdx}",
>  jsonLang: "**/*.{json,jsonc}"
> }
> ```
I think this makes sense. Would like to know the rest of the team's thoughts on this. @eslint/eslint-team

**Comment by JoshuaKGoldberg** [comment]
I'm a little biased here 🙂 but am very much in favor of this. As Kirk showed, it's been a real pain having users set up these globs.

For "vanilla" JS projects / ESLint core, most non-expert users I talk to don't understand when to use `.cjs` vs. `.js` vs. `.mjs`. Many aren't even aware of non-`.js` extensions that exist in their own (shared/team) codebases. Asking them to make reasonable decisions around file extensions is a lot. Doing this for them would lessen configuration confusion and help set up a precedent.

For userland extensions such as TypeScript, additional strong +1 on that precedent being helpful. https://github.com/typescript-eslint/typescript-eslint/issues/12102's suggested default extensions list contains _eight_ extensions -- each of which have legitimate reason to show up in any arbitrary TypeScript codebase. Having a precedent for plugins to include extension lists will be similarly helpful there.

**Comment by lumirlumir** [comment]
I also think this is a useful addition 👍, but I have two questions.

---

Is the `globs` helper intended for end users or for plugin authors?

If it's intended for end users, `@eslint/config-helpers` seems like a good place to land it. However, if it's intended for plugin authors or both, I'm not sure `@eslint/config-helpers` is the right location.

For example, [`eslint/markdown` defines `files: ["**/*.md"]`](https://github.com/eslint/markdown/blob/main/src/index.js#L94) and the plugin only depends on [`@eslint/core` and `@eslint/plugin-kit`](https://github.com/eslint/markdown/blob/main/package.json#L97-L98). If `globs` is implemented in `@eslint/config-helpers`, language plugins would need to add `@eslint/config-helpers` as a dependency, which seems undesirable.

So I'm not sure whether the best place is `@eslint/config-helpers`, `@eslint/plugin-kit`, or elsewhere.

---

Another question is: how broad should the language scope of `globs` be?

Currently, ESLint implements three language plugins: Markdown, JSON, and CSS. However, many community plugins now natively support language plugins, such as HTML and YAML.

So languages could include JavaScript, TypeScript, Markdown, JSON, CSS, HTML, and YAML, and `globs` might need to be expanded as new language plugins appear (e.g., Vue, TOML, Astro, etc.).

**Comment by kirkwaiblinger** [comment]
> Is the `globs` helper intended for end users or for plugin authors?

I filed this issue with the idea that it's for end users to be able to write configuration  files. If there's a use case for plugin authors I think that's more part of https://github.com/eslint/eslint/issues/19462

**Comment by kirkwaiblinger** [comment]
> Another question is: how broad should the language scope of `globs` be?
> 
> Currently, ESLint implements three language plugins: Markdown, JSON, and CSS. However, many community plugins now natively support language plugins, such as HTML and YAML.
> 
> So languages could include JavaScript, TypeScript, Markdown, JSON, CSS, HTML, and YAML, and `globs` might need to be expanded as new language plugins appear (e.g., Vue, TOML, Astro, etc.).

My criterion for success here would be that a user has a good way to know what glob(s) to reach for by default with a given rule. I'm not super familiar with the ESLint language system... 

- If languages are defined by ESLint are intended to be exhaustive/sealed, then a glob could be provided could apply to all files extensions within that language and adding new file extensions would be a breaking change (so this _might_ require ESLint to enumerate JS, TS, Vue, Astro, Svelte, etc, depending how the languages are defined in ESLint)
- If languages are not sealed, then it may be more appropriate to standardize a way for plugins to define globs, similar to how `plugins.configs` was standardized some time ago. This means though that the user might have to combine all plugins' globs in order to have rules apply to TS and Vue and Astro and Svelte, etc. Or, perhaps an automatic lookup/combination rule could leverage the standardized globs format, similar to what was done with `plugin.configs` 🤷‍♂️
- of course there may be another design I haven't thought of that fulfills this criterion much better! 🙂

**Comment by lumirlumir** [comment]
@kirkwaiblinger Thanks for the explanations! 

It would be nice if others could chime in and share their opinions.

**Comment by lumirlumir** [comment]
**TSC Summary**: This issue is seeking consensus on exporting common glob patterns for certain languages, such as `**/*.{mjs,cjs,js,jsx,mts,cts,ts,tsx}` for JavaScript.

**TSC Question**: There is currently no definitive consensus on the design of the glob patterns, where to place them, or how to treat them across the language plugins. Should we seek consensus on this approach, or do we need an RFC for this issue?

**Comment by nzakas** [comment]
We discussed this in the 2026-06-11 TSC Meeting. I'm not in favor of adding this to the core but I think the idea has merit as part of language plugins.

So `@eslint/js` could exports JS-related globs, `@eslint/markdown` could export Markdown-related globs, etc.

My suggestion would be to submit an RFC for extending language plugins with this information.

**Comment by kirkwaiblinger** [comment]
After reading #20999 I'm not sure that just exporting globs is the right approach. It helps people assemble their own final config when composing shared configs that need to be applied to a single language, but doesn't provide any help for utilizing shared configs containing rules that need to be applied to multiple different languages. This indicates that the right generalization would be aware of the ESLint `languages` system instead of being ad-hoc configured by the user with globs.
